src/Razor/src/Microsoft.AspNetCore.Razor.VSCode.Extension/syntaxes/aspnetcorerazor.tmLanguage.yml

```diff
@@ -103,47 +105,36 @@ repository:
     endCaptures:
       0: { name: 'punctuation.definition.tag.end.html' }
 
-  balanced-open-close-tag:
-    begin: (?=<([^/\s>]+))
-    patterns:
-      - include: '#balanced-open-close-tag-start'
-      - include: '#balanced-open-close-tag-end'
-      - include: '#balanced-open-close-tag-body'
-    end: (?<=(</\1>))|(/>)
-
-  balanced-open-close-tag-start:
-    begin: '(<)(?!(/))([^/\s>]+)'
-    beginCaptures:
-      1: { name: 'punctuation.definition.tag.begin.html'}
-      2: { name: 'constant.character.escape.razor.tagHelperOptOut' }
-      3: { name: 'entity.name.tag.html' }
-    patterns:
-      - include: '#balanced-open-close-tag-attributes'
-    end: '/?>'
+  # This is some complicated trickery to be able to detect & match
+  # self-closing components as well as elements with begin/end tags,
+  # more-or-less copied from microsoft/TypeScript-TmLanguage's TSX grammar.
+  # Basically - we match both ways a tag can be closed (self-closing or an end tag),
+  # and leave the inner patterns to figure out which one we matched.
+  non-void-tag:
+    begin: '(?=<(!)?[^/\s>]+(\s|/?>))'
+    end: '(/>)|(</)([^/\s>]+)\s*(>)'
```

Comment:
Ah, clever. I do like your approach here. I'd be curious to see how this all ends up impacting the  baselines. That being said, love that it's failing the CI. Wooo!

Reply:
> I do like your approach here

I can't take any credit, it's all thanks to the TypeScript language repo :). I didn't even think matching both self-closing tags and fully open/close tags was possible until looking to see how they did it for JSX!